src/Shared/NativeMethodsShared.cs

```diff
@@ -1181,17 +1245,76 @@ internal static int GetParentProcessId(int processId)
         /// Internal, optimized GetCurrentDirectory implementation that simply delegates to the native method
         /// </summary>
         /// <returns></returns>
-        internal static string GetCurrentDirectory()
+        internal unsafe static string GetCurrentDirectory()
         {
+#if FEATURE_LEGACY_GETCURRENTDIRECTORY
             if (IsWindows)
             {
-                StringBuilder sb = new StringBuilder(MAX_PATH);
-                int pathLength = GetCurrentDirectory(MAX_PATH, sb);
+                int bufferSize = GetCurrentDirectoryWin32(0, null);
+                char* buffer = stackalloc char[bufferSize];
+                int pathLength = GetCurrentDirectoryWin32(bufferSize, buffer);
+                return new string(buffer, startIndex: 0, length: pathLength);
+            }
+#endif
+            return Directory.GetCurrentDirectory();
+        }
+
+        private unsafe static int GetCurrentDirectoryWin32(int nBufferLength, char* lpBuffer)
+        {
+            int pathLength = GetCurrentDirectory(nBufferLength, lpBuffer);
+            VerifyThrowWin32Result(pathLength);
```

Comment:
I'm a bit surprised to see this here. Does it not attempt to throw when the first call (to compute `bufferSize`) returns nonzero?

Reply:
Not sure I understand, if the result is nonzero on either call then it continues without throwing.

Reply:
Actually I think I caused confusion by naming this one ```pathLength``` instead of say ```result```, this method is used in both calls above.

Reply:
Ah, you're right; I think I was confusing win32's error pattern with HRESULTs, which are clearer to me.